### PR TITLE
Signed packages

### PR DESCRIPTION
**Pull request by ilanschnell**
The purpose of this PR is to implement the ability to handle signed packages, ie. packages which have associated with them a signature file (`.sig`).  The signature is downloaded after the package itself has been downloaded and then verified against the public keys stored in the `.conda/keys` directory.

### CODE REVIEW

**Comment by ilanschnell** [comment]
See also https://github.com/conda/conda-build/pull/430

File: conda/signature.py [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+        res *= 256
+        res += (c if PY3 else ord(c))
+    return res
+
+
+def hash_file(path):
+    h = hashlib.new('sha256')
+    with open(path, 'rb') as fi:
+        while True:
+            chunk = fi.read(262144)
```
Comment (pzwang) [inline conda/signature.py]: Why this particular magic number in the call to `read()`?

File: conda/signature.py [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+def verify(path):
+    """
+    Verify the file `path`, with signature `path`.sig, against the key
+    found under ~/.conda/keys/<key_name>.pub
+    """
+    try:
+        with open(path + '.sig') as fi:
+            key_name, sig = fi.read().split()
+    except IOError:
+        return 'NO_SIGATURE'
```
Comment (pzwang) [inline conda/signature.py]: Is this function intended to only return one of three values "NO_SIGNATURE", "VALID", "INVALID"?  It seems to me that it's actually not very helpful to swallow all types of IOErrors and convert them into a single, opaque "NO_SIGNATURE" string.

If the rationale for swallowing exceptions is to make this function a simple utility that can be relied upon to return only one of three values, then this should also catch potential `ValueError`s resulting from the tuple unpacking of the call to `split()`.

File: conda/signature.py [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+    if key_name not in KEYS:
+        key_path = join(KEYS_DIR, '%s.pub' % key_name)
+        try:
+            KEYS[key_name] = RSA.importKey(open(key_path).read())
+        except IOError:
+            sys.exit("Error: no public key: %s" % key_path)
```
Comment (pzwang) [inline conda/signature.py]: Why trigger a `sys.exit()` here?  This makes it then impossible to reliably use this function (or any caller of this function) in other code that needs to rely upon conda as a library, because it could cause my program to exit.  (This also violates the design principle of having this `verify()` function only return one of three values, if indeed that was the design intent.)

File: conda/signature.py [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+from conda.compat import PY3
+
+from Crypto.PublicKey import RSA
+
+
+KEYS = {}
+KEYS_DIR = abspath(expanduser('~/.conda/keys'))
+
+
+def sig2ascii(i):
```
Comment (pzwang) [inline conda/signature.py]: Please write a docstring that defines, very accurately, what is a valid value for "i" here.  I would also suggest that the parameter should be called "sig" and not "i".

File: conda/signature.py [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+        i, r = divmod(i, 256)
+        ret.append(r)
+    if PY3:
+        s = bytes(n for n in ret[::-1])
+    else:
+        s = ''.join(chr(n) for n in ret[::-1])
+    return base64.b64encode(s).decode('utf-8')
+
+
+def ascii2sig(s):
```
Comment (pzwang) [inline conda/signature.py]: Please write a docstring that defines what is a valid input, and what an expected output should be.

File: tests/test_misc.py [hunk 6]
```diff
@@ -9,6 +10,16 @@ def test_cache_fn_url(self):
...
+    def test_ascii2sig(self):
+        self.assertEqual(sig2ascii(1234), 'BNI=')
+
+    def test_sig2ascii(self):
+        self.assertEqual(ascii2sig('BNI='), 1234)
+
+    def test_sig2ascii2sig(self):
+        for i in range(10000):
+            self.assertEqual(ascii2sig(sig2ascii(i)), i)
+
```
Comment (pzwang) [inline tests/test_misc.py]: Please also add corner case/out of bounds tests such as:
* negative values
* nulls/NaNs/zero
* numerical but invalid types (floats, decimal, datetime)

The code should handle these gracefully by raising an appropriate exception.  At least, the tests would verify that this code does not silently propagate invalid parameters and behavior under these circumstance.

(The reason I ask for these is because the concept of a "signature" has varied form and not everyone necessarily knows what to expect as the proper type of value (or range of values) for a signature.)

File: conda/fetch.py [hunk 7]
```diff
@@ -283,6 +283,17 @@ def fetch_pkg(info, dst_dir=None, session=None):
     path = join(dst_dir, fn)
 
     download(url, path, session=session, md5=info['md5'], urlstxt=True)
+    if info.get('sig'):
+        from conda.signature import verify
```
Comment (pzwang) [inline conda/fetch.py]: Why is this import inside the conditional block, and not at the top of the file?

File: conda/fetch.py [hunk 8]
```diff
@@ -283,6 +283,17 @@ def fetch_pkg(info, dst_dir=None, session=None):
...
+
+        fn2 = fn + '.sig'
+        url = (info['channel'] if info['sig'] == '.' else
```
Comment (pzwang) [inline conda/fetch.py]: Is `info['channel']` guaranteed to end in a trailing slash?  This code assumes that it does...

File: conda/fetch.py [hunk 9]
```diff
@@ -283,6 +283,17 @@ def fetch_pkg(info, dst_dir=None, session=None):
...
+               info['sig'].rstrip('/') + '/') + fn2
+        log.debug("url=%r" % url)
```
Comment (pzwang) [inline conda/fetch.py]: This is a duplicate of the `log.debug()` call in line 282 above.  That seems confusing when actually trying to debug this problem.  Maybe this should be `log.debug("signature url=%r" % url)`?

File: conda/fetch.py (hunk 7 above)
Comment (ilanschnell) [inline conda/fetch.py]: Because the signature module requires the pycrypto package, which I want to keep an optional dependency.

File: conda/fetch.py (hunk 8 above)
Comment (ilanschnell) [inline conda/fetch.py]: Yes, it is guaranteed to include the trailing '/'.

File: conda/fetch.py (hunk 9 above)
Comment (ilanschnell) [inline conda/fetch.py]: From the URL included in the output it is already clear that the file to be downloaded is a signature, but I also modified the debug message.

File: conda/signature.py (hunk 2 above)
Comment (asmeurer) [inline conda/signature.py]: I would make `"VALID"` -> `True`, `"INVALID"` -> `False`, and `"NO_SIGNATURE` (which is spelled wrong by the way) -> `None` (and be more careful with only catching the exception for file doesn't exist and let other IOErrors bubble up, as @pzwang noes).

File: conda/signature.py (hunk 1 above)
Comment (ilanschnell) [inline conda/signature.py]: This is actually not a magic number, it is just 2^18, ie. 256KB.  I added a comment.

File: conda/signature.py [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+    res = 0
+    for c in base64.b64decode(s):
+        res *= 256
+        res += (c if PY3 else ord(c))
+    return res
+
+
+def hash_file(path):
```
Comment (asmeurer) [inline conda/signature.py]: Wouldn't it be better to put this in `conda.utils` as `sha256_file`?

File: conda/fetch.py [hunk 11]
```diff
@@ -283,6 +283,17 @@ def fetch_pkg(info, dst_dir=None, session=None):
...
+        from conda.signature import verify
+
+        fn2 = fn + '.sig'
+        url = (info['channel'] if info['sig'] == '.' else
+               info['sig'].rstrip('/') + '/') + fn2
+        log.debug("signature url=%r" % url)
+        download(url, join(dst_dir, fn2), session=session)
+        v = verify(path)
+        if v != 'VALID':
+            sys.exit("Signature for '%s': %s" % (basename(path), v))
```
Comment (asmeurer) [inline conda/fetch.py]: This needs a better error message.

File: conda/signature.py (hunk 2 above)
Comment (ilanschnell) [inline conda/signature.py]: I agree, making those changes.

**Comment by ilanschnell** [comment]
I've improved the code, according to the very useful feedback.  I'm going to write some documentation now, which walks trough all the steps, from creating a repository with signed packages, to using them.

**Comment by ilanschnell** [comment]
Here is the documentation I wrote:
https://github.com/ContinuumIO/internal/blob/master/test-ilan/signed-packages.rst
